src/test/java/Hmwk21.java

```diff
@@ -1,2 +1,74 @@
-public class Hmwk21 {
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.chrome.ChromeOptions;
+import org.openqa.selenium.interactions.Actions;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.testng.annotations.Test;
+
+import java.util.concurrent.TimeUnit;
+
+
+//Using Action Class methods, create/refactor the test case renamePlaylist()
+//        Implement Explicit Waits
+//        Remove all Thread.sleep implementations
+//        Create a new branch and commit your changes
+//        Push your code to a remote repository
+//        Create a pull request
+//        Copy and paste the link of the pull request (or your branch) to the field below, so we can check your homework
+public class Hmwk21 extends BaseTest{
+
+    Actions action = new Actions(driver);
+
+
+@Test
+public void RenamePlaylist(){
+
+        ChromeOptions chrome_Profile = new ChromeOptions();
+        chrome_Profile.addArguments("--disable-notifications");
+        provideEmail("seamugg@yahoo.com");
+        providePassword("te$t$tudent");
+        clickSubmitBtn();
+        driver.manage().timeouts().implicitlyWait(5, TimeUnit.SECONDS);
+        GoToPlaylist();
+        driver.manage().timeouts().implicitlyWait(5, TimeUnit.SECONDS);
+        EditPlaylist();
+        AddNewName();
+
+
+}
+
+    private void EditPlaylist() {
+
+        WebElement EditPlaylist = driver.findElement(By.xpath("//button[contains(@title, 'Edit')]"));
+        action.moveToElement(EditPlaylist).click();
+
+    }
+
+
+    // i role="button" title="Create a new playlist" data-testid="sidebar-create-playlist-btn" class="fa fa-plus-circle create"></i>
+   // li data-v-5f33db96="" data-testid="playlist-context-menu-edit-27813">Edit</li>
+    private void AddNewName() {
+        WebElement AddNewName = driver.findElement(By.xpath("//li[contains(@testid, 'playlist-context-menu-edit-27813')]"));
+        AddNewName.click();
+        driver.manage().timeouts().implicitlyWait(5, TimeUnit.SECONDS);
+        AddNewName.sendKeys("NOT Playlist");
+
+    }
+
+//    private void CreateNewPlaylist() {
+//        WebElement renamePlaylist = driver.findElement(By.xpath("//button[contains(@title, 'Create a new playlist')]"));
+//        action.moveToElement(renamePlaylist).contextClick().perform();
+//    }
+//
+    public void GoToPlaylist() {
+        // WebElement GoToPlaylist= driver.findElement(By.xpath("//a[@href='#!/playlist/27538']"));
+        wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//a[@href='#!/playlist/27813']")));
+        WebElement GoToPlaylist= driver.findElement(By.xpath("//a[@href='#!/playlist/27813']"));
+        action.moveToElement(GoToPlaylist).contextClick().perform();
+
+
+    }
+
+
```

Comment:
I have corrected the BaseTest and the class NewNamePlayListTes:

1.  **BaseTest** 

```
public class BaseTest {

  protected WebDriver driver;
  private String url;

  @BeforeSuite
  public static void chromeConfigs() {
    if (System.getProperty("os.name").toLowerCase().contains("win")) {
      System.setProperty("webdriver.chrome.driver", "chromedriver.exe");
    }
  }

  @BeforeMethod
  public void launchBrowser() {
    ChromeOptions chromeOptions = new ChromeOptions();
    chromeOptions.addArguments("--disable-notifications");
    driver = new ChromeDriver(chromeOptions);
    driver.manage().window().maximize();
    driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(10));
    url = "https://bbb.testpro.io/#!/home";
    driver.get(url);
  }

  public void clickSubmitBtn() {
    WebElement submitButton = driver.findElement(By.cssSelector("[type='submit']"));
    submitButton.click();
  }

  public void provideEmail(String email) {
    WebElement emailField = driver.findElement(By.cssSelector("[type='email']"));
    emailField.click();
    emailField.sendKeys(email);
  }

  public void providePassword(String password) {
    WebElement passwordField = driver.findElement(By.cssSelector("[type='password']"));
    passwordField.click();
    passwordField.sendKeys(password);

  }

  protected WebElement waitForElementToBeClickable(WebElement webElementLocator) {
    return new WebDriverWait(driver, Duration.ofSeconds(10)).until(
        ExpectedConditions.elementToBeClickable(webElementLocator));
  }

  protected WebElement waitForVisibilityOfElement(WebElement webElementLocator) {
    return new WebDriverWait(driver, Duration.ofSeconds(10)).until(ExpectedConditions
        .visibilityOf(webElementLocator));
  }

  @AfterMethod
  public void tearDownBrowser() {
    driver.quit();
    driver = null;
  }

}
```

2. **NewNamePlayListTest**

```
public class NewNamePlayListTest extends BaseTest {
  String  newPlaylistName= "NOT Playlist";

  @Test
  public void RenamePlaylist() {

    provideEmail("seamugg@yahoo.com");
    providePassword("te$t$tudent");
    clickSubmitBtn();

    goToPlaylist();
    editPlaylist();
    addNewName();

  }

  private void editPlaylist() {
    Actions actions=new Actions(driver);
    WebElement editPlaylist = driver.findElement(By.xpath("//li[@data-testid='playlist-context-menu-edit-27813']"));
    actions.click(editPlaylist).perform();

  }

  private void addNewName() {
    WebElement addNewName = driver.findElement(By.xpath("//input[@name='name']"));
    waitForVisibilityOfElement(addNewName);
    addNewName.sendKeys((Keys.chord(Keys.CONTROL + "a", Keys.BACK_SPACE)));
    waitForVisibilityOfElement(addNewName);
    addNewName.sendKeys(newPlaylistName);
    addNewName.sendKeys(Keys.ENTER);

  }

  public void goToPlaylist() {
    WebElement goToPlaylist = driver.findElement(By.xpath("//a[@href='#!/playlist/27813']"));
    Actions actions=new Actions(driver);
    waitForElementToBeClickable(goToPlaylist);
    actions.click(goToPlaylist);
    actions.contextClick(goToPlaylist).perform();
  }
}
```

3. Please write **Assert.assertEquals**(.........);